src/fstext/kaldi-fst-io-inl.h

```diff
@@ -0,0 +1,220 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef KALDI_FSTEXT_KALDI_FST_IO_INL_H_
+#define KALDI_FSTEXT_KALDI_FST_IO_INL_H_
+
+#include "util/text-utils.h"
+
+namespace fst {
+
+
+template <class Arc>
+void WriteFstKaldi(std::ostream &os, bool binary,
+                   const VectorFst<Arc> &t) {
+  bool ok;
+  if (binary) {
+    // Binary-mode writing.  No binary header; the Read function
+    // knows it is text mode if it sees a space as the 1st character
+    // (the leading \n).
+    ok = t.Write(os, FstWriteOptions());
+  } else {
+    // Text-mode output.  Note: we expect that t.InputSymbols() and
+    // t.OutputSymbols() would always return NULL.  The corresponding input
+    // routine would not work if the FST actually had symbols attached.  Write a
+    // newline to start the FST; in a table, the first line of the FST will
+    // appear on its own line.
+    os << '\n';
+    bool acceptor = false, write_one = false;
+#ifdef HAVE_OPENFST_GE_10400
+    FstPrinter<Arc> printer(t, t.InputSymbols(), t.OutputSymbols(),
+                            NULL, acceptor, write_one, "\t");
+#else
+    FstPrinter<Arc> printer(t, t.InputSymbols(), t.OutputSymbols(),
+                            NULL, acceptor, write_one);
+#endif
+    printer.Print(&os, "<unknown>");
+    if (os.fail())
+      KALDI_ERR << "Stream failure detected writing FST to stream\n";
+    // Write another newline as a terminating character.  The read routine will
+    // detect this [this is a Kaldi mechanism, not something in the original
+    // OpenFst code].
+    os << '\n';
+    ok = os.good();
+  }
+  if (!ok) {
+    KALDI_ERR << "Error writing FST to stream";
+  }
+}
+
+// Utility function used in ReadFstKaldi
+template <class W>
+inline bool StrToWeight(const std::string &s, bool allow_zero, W *w) {
+  std::istringstream strm(s);
+  strm >> *w;
+  if (!strm || (!allow_zero && *w == W::Zero())) {
+    return false;
+  }
+  return true;
+}
+
+template <class Arc>
+void ReadFstKaldi(std::istream &is, bool binary,
+                  VectorFst<Arc> *fst) {
+  typedef typename Arc::Weight Weight;
+  typedef typename Arc::StateId StateId;
+  if (binary) {
+    // We don't have access to the filename here, so write [unknown].
+    VectorFst<Arc> *ans =
+        VectorFst<Arc>::Read(is, fst::FstReadOptions((std::string)"[unknown]"));
```

Comment:
C-style cast, syntactically. Better `std::string("...")`. Explicit constructions should be ok too I think.